From df5aa4e98dd4d1fed2f6c6ddf44dc759bd444033 Mon Sep 17 00:00:00 2001
From: Mostafa Saleh <smostafa@google.com>
Date: Thu, 16 Apr 2026 11:56:18 +0000
Subject: [PATCH 128/223] ANDROID: KVM: arm64: pviommu: Fix
 pkvm_guest_iommu_alloc_domain()

Fix error path in pkvm_guest_iommu_alloc_domain() where a spinlock was
released without getting acquired in the first place.

Bug: 503382913
Change-Id: I2ded996e408bb94554f92bd049b24aef2ff128d7
Signed-off-by: Mostafa Saleh <smostafa@google.com>
(cherry picked from commit f0a3950f0fa3988e9af43b5656f72fa675672c82)
Signed-off-by: Lee Jones <joneslee@google.com>
---
 arch/arm64/kvm/hyp/nvhe/iommu/pviommu.c | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/arch/arm64/kvm/hyp/nvhe/iommu/pviommu.c b/arch/arm64/kvm/hyp/nvhe/iommu/pviommu.c
index 7e7ea6628d5a..8358e6defac7 100644
--- a/arch/arm64/kvm/hyp/nvhe/iommu/pviommu.c
+++ b/arch/arm64/kvm/hyp/nvhe/iommu/pviommu.c
@@ -146,6 +146,11 @@ static bool pkvm_guest_iommu_alloc_domain(struct pkvm_hyp_vcpu *hyp_vcpu, u64 *e
 	struct kvm_hyp_req *req;
 	struct pkvm_hyp_vm *vm = pkvm_hyp_vcpu_to_hyp_vm(hyp_vcpu);
 
+	/* MBZ */
+	if (smccc_get_arg2(vcpu) || smccc_get_arg3(vcpu) || smccc_get_arg4(vcpu) ||
+	    smccc_get_arg5(vcpu) || smccc_get_arg6(vcpu))
+		goto out;
+
 	guest_domain = hyp_alloc(sizeof(*guest_domain));
 	if (!guest_domain) {
 		BUG_ON(hyp_alloc_errno() != -ENOMEM);
@@ -159,15 +164,10 @@ static bool pkvm_guest_iommu_alloc_domain(struct pkvm_hyp_vcpu *hyp_vcpu, u64 *e
 		return false;
 	}
 
-	/* MBZ */
-	if (smccc_get_arg2(vcpu) || smccc_get_arg3(vcpu) || smccc_get_arg4(vcpu) ||
-	    smccc_get_arg5(vcpu) || smccc_get_arg6(vcpu))
-		goto out_inval;
-
 	hyp_spin_lock(&pviommu_guest_domain_lock);
 	domain_id = pkvm_guest_iommu_alloc_id();
 	if (domain_id < 0)
-		goto out_inval;
+		goto out_unlock;
 
 	ret = kvm_iommu_alloc_domain(domain_id, KVM_IOMMU_DOMAIN_ANY_TYPE);
 	if (ret == -ENOMEM) {
@@ -178,7 +178,7 @@ static bool pkvm_guest_iommu_alloc_domain(struct pkvm_hyp_vcpu *hyp_vcpu, u64 *e
 		return false;
 	} else if (ret) {
 		pkvm_guest_iommu_free_id(domain_id);
-		goto out_inval;
+		goto out_unlock;
 	}
 
 	guest_domain->id = domain_id;
@@ -187,9 +187,10 @@ static bool pkvm_guest_iommu_alloc_domain(struct pkvm_hyp_vcpu *hyp_vcpu, u64 *e
 	smccc_set_retval(vcpu, SMCCC_RET_SUCCESS, domain_id, 0, 0);
 	return true;
 
-out_inval:
+out_unlock:
 	hyp_spin_unlock(&pviommu_guest_domain_lock);
 	hyp_free(guest_domain);
+out:
 	smccc_set_retval(vcpu, SMCCC_RET_INVALID_PARAMETER, 0, 0, 0);
 	return true;
 }
-- 
2.50.1

